### ISSUE TITLE
XGBoost implementation

### ISSUE THREAD

**Issue by Timo3as** [google/yggdrasil-decision-forests#13]
Hi all,

I'm trying to implement XGBoost, using gradient_boosted_trees with use_hessian_gain. After some tweaking of the parameter, while working with different datasets on binary classification problem, I cannot replicate the trees and results of xgboost's XGBClassifier; but as far as I understand the code, it should produce the same algorithm.

Did somebody tried this, or is this even possible? If yes can you point me in the direction of how I need to configure GradientBoostedTreesModel and if not is there an implementation of XGBoost planned?

Thanks so much in advance,

Timo

**Comment by achoum** [comment]
Hello Timo3as,
 
YDF and XGBoost are different but close implementations. With the right matching hyper-parameters, you can emulate more-or-less one of the libraries in the other.

When you say you cannot "replicate the trees", does it mean that you obtain different decision trees? If so, this is to be expected. Even within a given implementation, small changes (e.g. calling a random number generator in different orders) can lead to different models.

However, results should be / remain approximately similar (i.e. this is the training noise).

Therefore, with similar hyper-parameters (the names are not necessarily the same), you should at least obtain comparable results between TF-DF and XGBoost. If this is not the case, this could be an issue and this would be interesting for us to know :).

Cheers,
Mathieu

**Comment by Timo3as** [comment]
Hi again. We looked further into this matter, comparing the implemented algorithms, and I think we found a difference. To this time we are unsure, if this difference is of significance but wanted to share in case this information is useful. 

While comparing we were looking at the trees that the algorithms produced. We figured that trees in Yggdrasil were getting way larger than in XGBoost. In order to find out why, we compared the way the two algorithms are growing their trees and shed a light on how the algorithms decide if the split will happen.
We think we found the reason on why the trees grow larger.

To get a better view on the internals we used this small dataset:

Index   | Feature_a | Feature_b | Label
---     | ---       | ---       | ---
0       | 0         | 1         | 0
1       | 0         | 0         | 0
2       | 1         | 1         | 1
3       | 0         | 1         | 0

# Splits

## Xgboost split

### Split 1

Best split candidate:
- Feature: Feature_a
- Root Gain: 1
- Index Left: 2
- Index Right: 0, 1, 3
- Score Left: 1
- Score Right: 3
- Split happening: Yes

### Split 2

Index   | Feature_a | Feature_b | Label
---     | ---       | ---       | ---
0       | 0         | 1         | 0
1       | 0         | 0         | 0
2       | 0         | 1         | 0

Best split candidate:
- Feature: Feature_b
- Root Gain: 3
- Index Left: 0, 2
- Index Right: 1
- Score Left: 2
- Score Right: 1
- Split happening: No

## Yggdrasil split

### Split 1

Best split candidate:
- Feature: Feature_a
- Root Gain: 0
- Index Left: 2
- Index Right: 0, 1, 3
- Score Left: 1
- Score Right: 3
- Split happening: Yes

### Split 2

Index   | Feature_a | Feature_b | Label
---     | ---       | ---       | ---
0       | 0         | 1         | 0
1       | 0         | 0         | 0
2       | 0         | 1         | 0

Best split candidate:
- Feature: Feature_b
- <b>Root Gain: 0</b>
- Index Left: 0, 2
- Index Right: 1
- Score Left: 0.666667
- Score Right: 0.33333
- <b>Split happening: Yes</b>

## Explanation
Other than XGBoost Yggdrasil does not seem to carry through the gain of the previous split. It seems to get set to 0 before each new split is evaluated. 
We think this is why we found Yggdrasil trees to grow larger than XGBoost. Yggdrasil sets the score to 0 on every iteration, while XGBoost tries to find splits that get a better score than the one of the parent node.
The NodeCondition is initialized every time the [split function](https://github.com/google/yggdrasil-decision-forests/blob/1d0d6eb795a863732cd41ba75785434267c3136e/yggdrasil_decision_forests/learner/decision_tree/training.cc#L3126) is called.
And a new [NodeCondition](https://github.com/google/yggdrasil-decision-forests/blob/048ea31288e08e5f6a198fccb2b9c8f3906a4e41/yggdrasil_decision_forests/model/decision_tree/decision_tree.proto#L142) defaults to 0.
In this case this is no issue but it may or may not be an issue for more complex problems.

# Comparison

We tried to find some difference in performance on different datasets. These datasets were small toy datasets so far.

## Setup

XGBoost was running on MAC Os Laptop.
- 2,6 GHz 6-Core Intel Core i7
- 16 GB 2667 MHz DDR4
- AMD Radeon Pro 5300M 4 GB Intel UHD Graphics 630 1536 MB

TFDF was running von GoogleColab and Multipass.

Multipass:
- Ubuntu 20.04.3 LTS
- 14.4G Disk
- 3.8G Memory

## Datasets

### Sklearn - Breast Cancer Dataset

- Classes: 2
- Samples per class: 212 (M), 357 (B)
- Samples total: 569
- Dimensionality: 30
- Features: real, positive

### Sklearn - Digit Dataset

- Classes: 10
- Samples per class: 180
- Samples total: 1797
- Dimensionality: 64
- Features: integers 0-16

## Comparison

The datasets were loaded with sklearn and compared with different configurations of both models. We tried first to make them as equal as possible.

### Parameter

#### Yggdrasil

```
dt_kwargs_base = {
    'num_trees':100,
    'growing_strategy':"BEST_FIRST_GLOBAL",
    'max_depth':6,
    'use_hessian_gain':True,
    'sorting_strategy':"IN_NODE",
    'shrinkage':1.,
    'subsample':1.,
    'sampling_method': 'RANDOM',
    'l1_regularization':1.,
    'l2_regularization':1.,
    'l2_categorical_regularization':1.,
    'num_candidate_attributes': -1,
    'num_candidate_attributes_ratio': -1.,
    'min_examples':1,
    'validation_ratio':0.,
    'early_stopping':"NONE",
    'in_split_min_examples_check':False,
    'max_num_nodes': -1,
    'verbose': 0,
}
```

#### XGBoost

```
gb_kwargs = {
        'n_estimators': 1000,
        'max_depth': 6,
        'colsample_bytree': 1.,
        'colsample_bynode': 1.,
        'colsample_bylevel': 1.,
        'use_label_encoder': False,
        'reg_lambda': 1.,
        'reg_alpha': 1.,
        'min_child_weight': 0,
        'min_split_loss': 0,
        'max_delta_step': 0,
        'base_score': 0.5,
        'learning_rate': 1.,
        'tree_method': 'exact',
        'booster': 'gbtree',
        'nthread': 1,
        'eval_metric': eval_metric,
        'objective': objective,
        'subsample': 1,
        'verbosity': 0,
        'validate_parameters': False,
        'scale_pos_weight': 1,
        'refresh_leaf': 1,
        #'early_stopping_rounds': 10,
    }
```

Model/Dataset   | Breast Cancer | Digits 
---     | ---       | ---       
XGBoost       | 0.1072         | 0.1606         
Yggdrasil       | 0.0944         | 0.2350

# Conclusion

So far we are not sure if the different approaches of the algorithms is making a huge difference in performance of the model. After tweaking the hyperparameter Yggdrasil was able to perform equally as good or even better than XGBoost. So I wanted to reach out to you, if you can (un)verify our observation and give further insight.

Thanks a lot and best Regards

Timo
